### DISCUSSION TITLE
Koreader sync server

### DISCUSSION THREAD

**Discussion by Nathoufresh** [gotson/komga#1760] [Ideas]
Koreader sync server allows to synchronize the read progression of ebooks. It looks like Kavita implements it.

https://github.com/Kareadita/Kavita/discussions/2590

I think it would be a great addition to Komga.

Here's an implementation in Lua on Koreader's GitHub : https://github.com/koreader/koreader-sync-server

**Comment by gotson** [comment]
Funny enough i had a look at this last week for a friend.

There's a couple of pitfalls though:
- KOReader uses a hash to identify books, which can be:
  - a partial MD5 hash of the file itself. So if the file changes, the hash changes too.
  - a MD5 hash of the filename
- KOReader always logs in by sending a username in clear text, and a password in MD5 hash. Komga does not use MD5 (which is broken now), but BCrypt.

We would also need to know what to do with sync progress for books unknown to Komga. Should it be ignored ? Or should it be proxied to a KOReader Sync service (to be configured Komga side), so that it can still be stored/retrieved ?

Question for you: how do you acquire books on KOReader ? Do you use OPDS with the Komga OPDS feed ?

**Reply by gotson** [reply]
Note that if we do implement a proxy to another KOReader Sync server, we would need to store login/password per user inside Komga.

**Reply by Nathoufresh** [reply]
> KOReader uses a hash to identify books, which can be:
> - a partial MD5 hash of the file itself. So if the file changes, the hash changes too.
 > - a MD5 hash of the filename

Do you know what makes it uses the filename or the file itself? Does the behavior change with some koreader versions?

> KOReader always logs in by sending a username in clear text, and a password in MD5 hash. Komga does not use MD5 (which is broken now), but BCrypt.

Komga could generate a random username/password pair for the current user and display it in the parameters in clear. These would only be used for the sync progress server.

> We would also need to know what to do with sync progress for books unknown to Komga. Should it be ignored ? Or should it be proxied to a KOReader Sync service (to be configured Komga side), so that it can still be stored/retrieved ?

I guess it depends on the use case, for mine it could just be ignored 😋

> how do you acquire books on KOReader ? Do you use OPDS with the Komga OPDS feed ?

Yes exactly.

**Reply by gotson** [reply]
> Do you know what makes it uses the filename or the file itself? Does the behavior change with some koreader versions?

It's a choice from the user in KOReader

**Comment by UbadahJ** [comment]
This feature request sounds like something I’d really find useful. Is this being considered, or is there any update on it?

Also, are there any APIs I could use to get or update progress related to books? If not, I’m thinking of setting up a temporary proxy server to handle it and manage auth manually. Would that be a workable approach?

**Comment by gotson** [comment]
🎉 This issue has been resolved in `1.17.0` ([Release Notes](https://github.com/gotson/komga/releases/tag/1.17.0))
